### ISSUE TITLE
Weird rotation glitch on Node2D with negative scale values

### ISSUE THREAD

**Issue by MightyPrinny** [godotengine/godot#29625]
**Godot version:**
3.1.1 stable

**Issue description:**
1- If an object has a rotation of 180 degrees in the editor rotation_degrees will be 179.999847
 after the game starts

2-A very weird glitch happens if you change global_position and then change global_transform to the same value when the node has negative scale values that causes the node to rotate constantly

**Steps to reproduce:**
this script is enough to reproduce it

```
extends Node2D
func _physics_process(delta):
	do_stuff()
func do_stuff():
	var xmotion =Vector2(0,0) 
	var ymotion =Vector2(0,0)
	global_position += xmotion
	global_position += ymotion
	more_stuff()
func more_stuff():
	var pre_trans = global_transform
	global_transform = pre_trans
```

**Minimal reproduction project:**
[RotationBugs.zip](https://github.com/godotengine/godot/files/3269395/RotationBugs.zip)

**Comment by Zireael07** [comment]
Point 1 is simply floating point accuracy, it's normal.

**Comment by MightyPrinny** [comment]
I find it weird because the engine actually handles this very well in other situations, vectors rotated by for example deg2rad(90) have very accurate values, in fact the rotation doesn't have imprecision .

**Comment by MightyPrinny** [comment]
@Zireael07  and this code prints 180 twice so rotation_degrees can be 180
```
rotation_degrees = 180
print(rotation_degrees)
rotation = rotation
print(rotation_degrees)
```

**Comment by groud** [comment]
Floating point inaccuracies are not an issue. When the 2D transform of an object is updated, this might lead to such inaccuracies, which are an expected behavior.
The second point is valid though.

**Comment by avencherus** [comment]
May be related to an older issue I posted:  https://github.com/godotengine/godot/issues/21020

For me a -1x scaling apparently was being globally represented as -1y and 180 degree rotation.  Though sure, identical, it is a bit unexpected and unusual to deal with in GDScript.

**Comment by MightyPrinny** [comment]
I don't think this specific inaccuracy is an expected behavior, the engine should be consistent, this does not happen if you set the rotation to 90, 270 or 0

**Comment by aaronfranke** [comment]
1 is caused by the rotation being saved as `3.14159` in the scene file, which is 6 digits (the number of reliable float digits). Ideally numbers should be saved in `tscn` files, not just with the reliable digits, but also with the unreliable digits, so reading a saved value will give the same value.

(Note that floating-point error still exists, so the closest we can get with singles is `180.000005` degrees, but that's still much closer than `179.999847`).

2 is fixed by #39611, the existing code is bogus.

**Comment by akien-mga** [comment]
Closing as fixed by #39611.

The first point is not a bug per se, but if you think we should reconsider the way floats are serialized, that could be discussed in a proposal on https://github.com/godotengine/godot-proposals. But note that serializing unreliable digits might mean that `.tscn` files will see arbitrary changes based on which platforms they are saved in, which could be a bother in collaborative projects.

**Comment by aaronfranke** [comment]
In case solving the first point is desired, [here's a patch file](https://github.com/godotengine/godot/files/4794011/0001-Make-float-also-write-unreliable-digits-to-tscn-file.txt) to write unreliable digits for float values. It doesn't work for things that aren't Variant::FLOAT, for example, vectors, so it's not good as-is. If this is added, it needs to be demonstrated that this is desired and works consistently across platforms.

Regardless, it's important to accept that checking if it's equal to 180 will never work. Try this instead:

    print(is_equal_approx(179.999847, 180)) # Prints "True"
